### PR TITLE
[feat]: [CDS-37734]: Implement deployStack for NG

### PR DESCRIPTION
**Pull request by nassergonzalez**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)

### CODE REVIEW

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java` line 61 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2LmoKv3v6V1qCn1CP0:-N2LmoKw8mmjE02HTGtQ:b-8wq37g) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java#L61)):*
> ```Java
>       CloudformationTaskNGParameters taskNGParameters, String delegateId, String taskId, LogCallback logCallback)
>       throws IOException, TimeoutException, InterruptedException {
>     remainingTimeoutMs = taskNGParameters.getTimeoutInMs() > 0
> ```

is there a scenario when timeout s coming < 0?



_Code quote:_
```Java
long endTime
```
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java` line 310 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2LrHKE15Klrm2EPia4:-N2LrHKE15Klrm2EPia5:b5o8eoo) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java#L310)):*
> ```Java
>     DeployStackResult deployStackResult = awsCloudformationClient.deployStack(
>         request.getRegion(), deployStackRequest, awsInternalConfig, Duration.ofMillis(endTime), logCallback);
>     if (deployStackResult.getStatus() == Status.FAILURE) {
> ```

can we have here the same logic as in CG?
``` if (deployStackResult.getStatus() == Status.FAILURE) {
      throw new InvalidRequestException(
          format("CF Change set creation failed: %s", deployStackResult.getStatusReason()));
    }
```
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/cloudformation/CloudformationTaskNGResponse.java` line 29 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2LgrU12DNGRGUruqf6:-N2LgrU12DNGRGUruqf7:b-ckmno1) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/cloudformation/CloudformationTaskNGResponse.java#L29)):*
> ```Java
>   CloudFormationCommandNGResponse cloudFormationCommandNGResponse;
>   CommandExecutionStatus commandExecutionStatus;
>   boolean stackSkipped;
> ```

maybe stackUpdated?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java [hunk 1]
```diff
@@ -292,20 +303,20 @@ private void updateStackAndWaitWithEvents(CloudformationTaskNGParameters request
...
-    UpdateStackResult updateStackResult =
-        awsCloudformationClient.updateStack(request.getRegion(), updateStackRequest, awsInternalConfig);
+    DeployStackRequest deployStackRequest = cloudformationBaseHelper.transformToDeployStackRequest(updateStackRequest);
+    long endTime = remainingTimeoutMs + System.currentTimeMillis();
+    DeployStackResult deployStackResult = awsCloudformationClient.deployStack(
+        request.getRegion(), deployStackRequest, awsInternalConfig, Duration.ofMillis(endTime), logCallback);
+    if (deployStackResult.getStatus() == Status.FAILURE) {
+      String errorMessage = format("# Error creating changeSet: %s", deployStackResult.getStatusReason());
+      logCallback.saveExecutionLog(errorMessage, LogLevel.ERROR);
+      populateCloudformationTaskNGResponse(builder, null, false);
```
Comment (vlica-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java:313]: I think that when we do an update, we always have an existing stack, even if creation of changeSet failed for any reason.

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/cloudformation/CloudformationCreateStackStep.java` line 172 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2M5OMw3MPhKTsb9QVT:-N2M5OMw3MPhKTsb9QVU:bwhlyto) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/125-cd-nextgen/src/main/java/io/harness/cdng/provision/cloudformation/CloudformationCreateStackStep.java#L172)):*
> ```Java
>         (CloudFormationCreateStackNGResponse) cloudformationTaskNGResponse.getCloudFormationCommandNGResponse();
> 
>     if (!cloudformationTaskNGResponse.isStackSkipped()) {
> ```

we should always save cloudformationIngeritOutput it is use also when we want to use Delete stack inherited.
We shouldn't save cloudformationConfig if there was no update

**Review by nassergonzalez — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 1 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/cloudformation/CloudformationCreateStackStep.java` line 172 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2M5OMw3MPhKTsb9QVT:-N2MPPd7AQif7Pf-zemG:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/125-cd-nextgen/src/main/java/io/harness/cdng/provision/cloudformation/CloudformationCreateStackStep.java#L172)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

we should always save cloudformationIngeritOutput it is use also when we want to use Delete stack inherited.
We shouldn't save cloudformationConfig if there was no update
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java` line 61 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2LmoKv3v6V1qCn1CP0:-N2MFoAa41j9UDRBeNAG:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java#L61)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

is there a scenario when timeout s coming < 0?
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java` line 310 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2LrHKE15Klrm2EPia4:-N2MG2tq8VE6ph4hjaeB:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java#L310)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

can we have here the same logic as in CG?
``` if (deployStackResult.getStatus() == Status.FAILURE) {
      throw new InvalidRequestException(
          format("CF Change set creation failed: %s", deployStackResult.getStatusReason()));
    }
```
</blockquote></details>

Done.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java` line 313 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2Le4-erMBFO_1o97-7-r1-313:-N2MG3MYCEz6WEDtdj3C:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/930-delegate-tasks/src/main/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandler.java#L313)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

Regarding the third parameter (false). I think that when we do an update, we always have an existing stack, even if creation of changeSet failed for any reason.
</blockquote></details>

Done.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/cloudformation/CloudformationTaskNGResponse.java` line 29 at r1](https://reviewable.io/reviews/harness/harness-core/33303#-N2LgrU12DNGRGUruqf6:-N2MPO2T15V9ObCnfBpL:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c0a46761f5765924351794d58641d5a81784c1ca/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/cloudformation/CloudformationTaskNGResponse.java#L29)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

maybe stackUpdated?
</blockquote></details>

Done.

File: 930-delegate-tasks/src/test/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandlerTest.java [hunk 2]
```diff
@@ -298,13 +306,84 @@ public void testUpdateNewStackIsFailureCauseOfException() throws IOException, In
...
+    assertThat(response.getCommandExecutionStatus()).isEqualTo(FAILURE);
+    verify(cloudformationBaseHelper, times(1)).getCloudformationTags(anyString());
+    verify(cloudformationBaseHelper, times(1)).getCapabilities(any(), anyString(), anyString(), any(), any());
+    verify(awsCloudformationClient, times(1)).deployStack(anyString(), any(), any(), any(), any());
+    verify(awsCloudformationClient, times(1)).getAllStacks(anyString(), any(), any());
+    verify(cloudformationBaseHelper, times(0)).printStackEvents(any(), anyString(), anyLong(), any(), any());
+    verify(cloudformationBaseHelper, times(0)).printStackResources(any(), anyString(), any(), any());
+    ArgumentCaptor<String> logCaptor = ArgumentCaptor.forClass(String.class);
+    verify(logCallback, atLeastOnce()).saveExecutionLog(logCaptor.capture());
+    assertThat(logCaptor.getAllValues().contains("# Update Successful for stack"));
```
Comment (vlica-harness) [inline 930-delegate-tasks/src/test/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandlerTest.java:348]: Hi Nasser, 
It looks strange why we get this log "Update Successful for stack", when we test/assert for failure update.

**Review by tmacari-harness — COMMENTED** [review]
> Reviewed 3 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 1 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`930-delegate-tasks/src/test/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandlerTest.java` line 348 at r2](https://reviewable.io/reviews/harness/harness-core/33303#-N2Le4-erMBFO_1o97-8-r2-348:-N2QPCU_CEs4YqlAS2LQ:bh29xjf) ([raw file](https://github.com/harness/harness-core/blob/5a2751759adc0a9b144c105a1836777fd745cedf/930-delegate-tasks/src/test/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandlerTest.java#L348)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

Hi Nasser, 
It looks strange why we get this log "Update Successful for stack", when we test/assert for failed (FAILURE) update.
</blockquote></details>

Good catch. It should be 
"assertThat(logCaptor.getAllValues()).contains("# Update Successful for stack");"
and probably it would fail.

Please fix in other places in that test class.

**Review by nassergonzalez — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 1 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`930-delegate-tasks/src/test/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandlerTest.java` line 348 at r2](https://reviewable.io/reviews/harness/harness-core/33303#-N2Le4-erMBFO_1o97-8-r2-348:-N2QhADH9PXf9MNTkO1c:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/5a2751759adc0a9b144c105a1836777fd745cedf/930-delegate-tasks/src/test/java/io/harness/delegate/task/cloudformation/handlers/CloudformationCreateStackTaskHandlerTest.java#L348)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

Good catch. It should be 
"assertThat(logCaptor.getAllValues()).contains("# Update Successful for stack");"
and probably it would fail.

Please fix in other places in that test class.
</blockquote></details>

Done.

**Review by vlica-harness — DISMISSED** [review]

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 2. unresolved discussions.

**Review by vlica-harness — APPROVED** [review]

**Review by vlica-harness — DISMISSED** [review]

**Review by abhinav-harness — DISMISSED** [review]
> Reviewed 2 of 4 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 2. unresolved discussions.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303)* status: 2. unresolved discussions.

**Review by nassergonzalez — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33303#-:-N2Vtaxq7wcrYTBdbWTH:bnfp4nl)* status: 2. unresolved discussions.

**Pull request merged**